Hypervisor (H) extension for RISCV

Navyera opened the pull request:
This set of commits provides an initial implementation of the H-extension for RISCV cores in gem5.

The implementation seems functional for the AtomicCPU, however more changes are needed for O3CPU to function correctly in H-extension scenarios. We plan to add these changes in a follow-up PR possibly.

In summary the implemented features include:

- New CSRs mandated by the H-extension
- CSR write mask support (CSRExecute rework)
- Additional fault handling & delegation
- Extensions for existing instruction implementations
- G-Stage translation support
- Two-stage translation support
- New instructions mandated by the H-extension

Navyera committed:
arch-riscv: Registers and Read Masks for H-extension

- Added 'MHSU' privilege mode set which implicitly enables the H-extension
- Added MISCREGs and CSRs for H-extension
- Added offsets for new registers' fields

Change-Id: I80789a78264dcc22f044a9532e629aaa3859c601

Navyera committed:
arch-riscv: Fault codes and types for H-extension

- Refactored fault codes enumerations into fault_codes.hh
- Added H-extension interrupt register fields (misc.hh)
- Extended fault codes for H-extension
- Extended fault classes for H-extension

Change-Id: If2d693150c5f18fee0aecef40ddca616026c06dd

Navyera committed:
arch-riscv: V-bit utils and FPU, VPU status update

- Added utilities for checking, setting and resetting the V-bit (H-extension)
- Refactored FPU and VPU checks into small reusable functions in isa.cc
- Refactor all float and vector op checks to use the new functions

Change-Id: I03255635910eea89af9deb350f72afd995f7f58c

Navyera committed:
arch-riscv: CSR operations refactor and additions for H-extension

- Write masks for CSRs (not all visible bits are writable)
- Check access to HPM CSRs according to spec
- Re-implement CSRExecute with better abstractions
	- readCSR (applies readmask)
	- writeCSR (applies writemask and respects physical register bits that are not visible)
        - backdoorReadCSRAllBits (get all bits of the physreg the CSR is mapped to)
- Revert changes that break the separation of physical registers of our "hardware" (MISCREG) with ISA specified constructs (CSR)

Change-Id: Ia9df3d20af4b1e20db49c081bda8b546752da418

Navyera committed:
arch-riscv: Fault handling and delegation for H-extension

- Extend globalMask() creation (interrupts.hh)
- Extend RiscvFault::invoke to handle and delegate correctly

Change-Id: I405a0423ff3cb39e059e08725d0128f7c3e7379c

Navyera committed:
arch-riscv: Extend existing instructions for H-extension

- Small fixes for previous commits to compile
- Extend mret, sret to handle transitions with new privileges
- Throw virtual faults where needed

Change-Id: Iee9d84fc7286e85f29594be6875cc160d5c2ad8f

Navyera committed:
arch-riscv: Implement two-stage translation for AtomicCPU

- Implement GStage walks
- Implement TwoStage walks
- Walks are of three types now:
	- FIRST_STAGE
	- GSTAGE_ONLY
	- TWO_STAGE
- Extend permission checking for new walks
- Extend pagefault creation for new fault types
- Store guest pte in TlbEntry to correctly check permissions for some cases
- Implement Walker::WalkerState::debug_condition to filter which walks are logged when debugging with PageTableWalker flag on
- Reserved more ARCH_BITS to signal:
	- Forced virtualization
	- HLVX
	- LR

Change-Id: I0801676f610e796430c61c6489999700ed8ae992

Navyera committed:
arch-riscv: Implement new H-extension instructions

- Hypervisor Loads, Stores, Fences, etc.
- Small change to enable H-extension in RiscvISA.py by default

Change-Id: I033ea064b904dfcb6bcff2a2dc0424113e47e245

rogerchang23424 commented on src/arch/riscv/isa/formats/fp.isa:
```diff
@@ -36,20 +36,20 @@ def template FloatExecute {{
         %(op_rd)s;
 
+        RegVal FFLAGS = xc->readMiscReg(MISCREG_FFLAGS);
         std::feclearexcept(FE_ALL_EXCEPT);
         %(code)s;
-        xc->setMiscReg(MISCREG_FFLAGS_EXE, softfloat_exceptionFlags);
+
+        FFLAGS |= softfloat_exceptionFlags;
         softfloat_exceptionFlags = 0;
+        xc->setMiscReg(MISCREG_FFLAGS, FFLAGS & FFLAGS_MASK);
```
If you do the change again. It will cause fflags issue in O3CPU https://github.com/gem5/gem5/issues/755

Navyera replied:
Hey, thanks for the comment, we had doubts about that part. Could you briefly explain what the bug is here? I read #755  but from the traces given I do not yet understand the "logic" behind the bug. Should we set FFLAGS instead of OR'ing?

rogerchang23424 replied:
It means we will get the wrong fflags result in O3CPU compared with AtomicCPU, you can test via command

```bash
build/RISCV/gem5.opt --debug-flags=Exec tests/gem5/configs/simple_binary_run.py rv64uf-ps-fadd o3 riscv --num-cores 4 --resource-directory tests/gem5/resources
```

For atomic CPU
```bash
build/RISCV/gem5.opt --debug-flags=Exec tests/gem5/configs/simple_binary_run.py rv64uf-ps-fadd atomic riscv --num-cores 4 --resource-directory tests/gem5/resources
```
> Should we set FFLAGS instead of OR'ing?

No, we accumulate the FFLAGS in each float arithmetic operation.

It's because the O3CPU[3] handles setMiscReg in different way compare to other CPU models, SimpleCPU[1] and MinorCPU[2]. The O3CPU wouldn't write the value back to MiscRegFile until the instruction is committed, we may get the old FFLAGS value in the next float instruction, and cause the riscv-tests failed. So that PR https://github.com/gem5/gem5/pull/868  use eatra misc_reg index `MISCREG_FFLAGS_EXE` to accumulate the FFLAGS to `MISCREG_FFLAGS` `miscRegFile`, just like what is implemented in the ARM[4]. ARM handles each float arithmetic exception via `MISCREG_FPSCR_EXC` rather than read/write `MISCREG_FPSCR` directly.

[1] SimpleCPU
https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/simple/exec_context.hh#L259-L263

[2] MinorCPU
https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/minor/exec_context.hh#L263-L267

[3] O3CPU
https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/o3/dyn_inst.hh#L1016-L1032

[4] 
https://github.com/gem5/gem5/blob/47c4dad8694448c40435875309a8399abebd7fc2/src/arch/arm/isa.cc#L890-L896

I seperated it with `NUM_PHYS_MISCREGS` and `NUM_MISCREGS` just because there is no reason to upgrade the checkpoint and avoid the checkpoint breakage.

Navyera committed:
arch-riscv: fix compiler warnings

Change-Id: I412b1b70b7f2a8b64953996636c7d59c5ec75581

BobbyRBruce commented:
The RISCV tests are failing with:

```
src/arch/riscv/pagetable_walker.cc:454: panic: panic condition misa.rvh occurred: Timing walks are not supported with h extension
```

hnpl commented on src/arch/riscv/RiscvISA.py:
```diff
@@ -106,7 +107,9 @@ class RiscvISA(BaseISA):
     )
     privilege_mode_set = Param.PrivilegeModeSet(
-        "MSU",
+        "MHSU",  # MHSU to enable hypervisor (H-extension)
```
The MHSU mode should not be the default if the timing cpu/timing mode is not supported.

Navyera replied:
Yeah that is totally reasonable, I just wanted to see how many of the tests were passing with H on. If H-extension is disabled a lot of logic is guarded out. 

I am going to set the default to MSU and push, hopefully tests will pass :)

powerjg commented:
Thanks for the contribution! I'm excited to dig in to review the change, and I'm excited for this feature to be merged!

Navyera committed:
arch-riscv: Set default to MSU for functional O3CPU

Change-Id: I263cda2eba8266c4f516e9c06845325b761830a0

rogerchang23424 commented on src/arch/riscv/regs/misc.hh:
```diff
@@ -200,6 +201,42 @@ enum MiscRegIndex
     MISCREG_VTYPE,
     MISCREG_VLENB,
 
+    // H-extension registers (RV64)
+
+    MISCREG_HVIP, // vssi bit is alias for mip.vssi
+
+    MISCREG_MTINST,
+    MISCREG_MTVAL2,
+
+    MISCREG_HSTATUS,
+    MISCREG_HEDELEG,
+    MISCREG_HIDELEG,
+    MISCREG_HCOUNTEREN,
+    MISCREG_HGEIE,
+    MISCREG_HTVAL,
+    MISCREG_HTINST,
+    MISCREG_HGEIP,
+
+    MISCREG_HENVCFG,
+    MISCREG_HGATP,
+    MISCREG_HCONTEXT,
+    MISCREG_HTIMEDELTA,
+
+    MISCREG_VSSTATUS,
+    MISCREG_VSTVEC,
+    MISCREG_VSSCRATCH,
+    MISCREG_VSEPC,
+    MISCREG_VSCAUSE,
+    MISCREG_VSTVAL,
+    MISCREG_VSATP,
+
+    // This register stores the V-bit.
+    // It is not mandated by the specification,
+    // however it is useful for the implementation.
+    MISCREG_VIRT,
+
+    // H-extension registers end here
```
You may need to update the checkpoint upgrade `util/cpt_upgraders`

Navyera replied:
We have locally updated the cpt_upgrader but we have this failing test: test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt.

The code tries to restore some of the new MISCREGs for H-extension and crashes.

Do we have to upgrade the checkpoint beforehand? Would the code automatically try to do that?

rogerchang23424 commented on src/arch/riscv/isa.cc:
```diff
@@ -984,6 +1003,250 @@ ISA::resetThread()
+
+
+    switch (csr) {
+        case CSR_FCSR: {
+            setMiscReg(MISCREG_FFLAGS, bits(writeDataMasked, 4, 0));
+            setMiscReg(MISCREG_FRM, bits(writeDataMasked, 7, 5));
+            break;
+        }
```
(nit) I hope we can align the implement of `MISCREG_VCSR`. They are do similar things.

https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/arch/riscv/isa.cc#L595-L599

rogerchang23424 commented on src/arch/riscv/isa/formats/standard.isa:
```diff
@@ -339,7 +339,7 @@ def template CSRExecute {{
 
         RegIndex midx = csr_data_it->second.physIndex;
-        const std::string& csrName = csr_data_it->second.name;
+        std::string csrName = csr_data_it->second.name;
```
(nit) Is any reason to change it to copy the string from csr_data to csrName?

Navyera replied:
It can no longer be const or reference due to possible remapping (reassignment) by swapToVirtCSR method.

rogerchang23424 commented on src/arch/riscv/isa/formats/standard.isa:
```diff
@@ -394,7 +433,7 @@ def template CSRExecute {{
             DPRINTF(RiscvMisc, "Writing %#x to CSR %s.\n",
                     data, csrName);
-            xc->setMiscReg(midx, data);
+            isa->writeCSR(csr, data);
```
It may affect the statistics of MiscReg if the `writeCSR` doesn't set miscReg value via `xc->setMiscReg`

https://github.com/gem5/gem5/blob/d2c8754ab3b72e52a07d62662de0120b941f1ae5/src/cpu/simple/exec_context.hh#L259-L263

Navyera replied:
I think all paths of writeCSR result in setMiscReg being called

powerjg replied:
How can you be sure of this since this line changes from `setMiscReg` to `writeCSR`?

powerjg replied:
To test this, you can run something like Linux boot (riscv-ubuntu-boot workload) on top of the "base" develop and with your changes and compare the read/write miscregs numbers in the stats.

Navyera replied:
@rogerchang23424  was totally right on this one, we changed readCSR/writeCSR to use xc->read/write so that miscReg  stats are updated

rogerchang23424 reviewed (commented):
I don't look into the details of hypervisor spec. I just do review briefly

rogerchang23424 commented on src/arch/riscv/isa.cc:
```diff
@@ -984,6 +1003,250 @@ ISA::resetThread()
+    auto csr_it = CSRData.find(csr);
+
+    // panic if the method was used with bad csr idx
+    panic_if(csr_it == CSRData.end(), "Illegal CSR passed to readCSR");
```
Raise `IllegalInst` instead of `panic_if`?

Navyera replied:
I think it has been checked at the start of CSR execute (standard.isa:335).

```
auto csr_data_it = csr_data.find(csr);
if (csr_data_it == csr_data.end()) {
    return std::make_shared<IllegalInstFault>(
        csprintf("Illegal CSR index %#x\n", csr), machInst);
}
```

So the panic_if in readCSR is just defensive programming.

ivanaamit commented:
Hi @Navyera, could you please address Roger's comments? Thank you.

Navyera commented:
Hi @ivanaamit , @rogerchang23424. I am sorry for the late reply, I was on summer vacation, but I'm now back and looking into the comments and test issues. I'll provide an update as soon as possible.

Navyera committed:
arch-riscv: Include FFLAGS O3 CPU fix from #868

Change-Id: Iff56e9f92bc2fa6a0d2be5de3c1b87dd8cf59074

Navyera commented:
Hi @rogerchang23424 @ivanaamit , we have addressed all of the comments hopefully.

Currently the only failing test is `test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt` which probably uses a precomputed checkpoint as a resource.

How should we go about updating that resource to include default values for the new H-extension registers we have added?

Harshil2107 commented:
Hi, to update the checkpoint you would probably need to update the `checkpoint_upgrader` in `util/cpt-upgrader` and `util/cpt-upgrader.py`. I can then update the checkpoint and upload the new one to gem5 resources.

Navyera committed:
arch-riscv: Fix SIE write mask

Change-Id: I7cd623dcb9529cf27c0da250814b0f4e0f6bcb8d

Navyera committed:
arch-riscv: Add riscv_hext cpt upgrader

Change-Id: Ia5f2a7125bc4b47a10196e8bc5d836eb2693ff55

Navyera commented:
Hi @Harshil2107 , I have updated the checkpoint upgrader and in my local testing, checkpoints seem to be succesfully upgraded. Let me know if you encounter any weirdness with the `riscv-hext.py` upgrader!

Navyera commented:
Hi @Harshil2107 @ivanaamit , the last failed test run has the same issue with the checkpoint that is not upgraded for the H-extension.

We have already added the relevant upgrader, but the resource does not seem to be updated yet.

On local runs I am getting the following:

```
command line: ./build/ALL/gem5.opt ./configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py

info: Using default config
Resource 'riscv-hello-example-checkpoint' was not found locally. Downloading to '/home/nick/.cache/gem5/riscv-hello-example-checkpoint.tar'...
Finished downloading resource 'riscv-hello-example-checkpoint'.
Unpacking the the resource 'riscv-hello-example-checkpoint' ('/home/nick/.cache/gem5/riscv-hello-example-checkpoint.tar')
Global frequency set at 1000000000000 ticks per second
src/mem/dram_interface.cc:690: warn: DRAM device capacity (8192 Mbytes) does not match the address range assigned (32 Mbytes)
src/arch/riscv/isa.cc:311: info: RVV enabled, VLEN = 256 bits, ELEN = 64 bits
src/base/statistics.hh:279: warn: One of the stats is a legacy stat. Legacy stat is a stat that does not belong to any statistics::Group. Legacy stat is deprecated.
board.remote_gdb: Listening for connections on port 7000
src/sim/globals.cc:91: warn: **********************************************************
src/sim/globals.cc:92: warn: !!! Checkpoint is missing the following version tags:
src/sim/globals.cc:96: warn:   riscv-hext
src/sim/globals.cc:96: warn:   x86-add-xcr0
src/sim/globals.cc:100: warn: You might experience some issues when restoring and should run the checkpoint upgrader (util/cpt_upgrader.py) on your checkpoint
src/sim/globals.cc:103: warn: **********************************************************
src/sim/serialize.hh:379: fatal: fatal condition !paramInImpl(cp, name, param) occurred: Can't unserialize 'board.processor.cores.core.xc.0:hvip_ulong'
Memory Usage: 278596 KBytes

```

Harshil2107 commented:
Hi @Navyera ,

Sorry for the delay, I didn't get the notification of the original ping. I will run the upgrader on the checkpoints on resources and update them.

Harshil2107 commented:
I am trying to restore the following checkpoint https://resources.gem5.org/resources/riscv-hello-example-checkpoint/raw?database=gem5-resources&version=3.0.0 but it seems like it is missing some keys. I just wanted to verify if this was the checkpoint you were able to upgrade locally. This is the output I get when trying to restore the checkpoint.

I just downloaded the checkpoint from the `url`and untarred it, then ran the upgrader.
```bash
python3 ./util/cpt_upgrader.py -v  ./riscv-checkpoint-v4/
Processing file ./riscv-checkpoint-v4/m5.cpt.... 

has tags riscv-vext memory-per-range arm-sysreg-mapping-ns armv8 riscv-dyn-vlen mempool-to-seworkload multiple-event-queues smt-interrupts arm-ccregs dvfs-perflevel arm-sve cpu-pid arm-gicv2-banked-regs x86-add-tlb etherswitch globals-to-root riscv-pcstate arm-miscreg-teehbr register-files ide-dma-abort arm-hdlcd-upgrade ruby-block-size-bytes process-fdmap-rename mempool-sections arm-contextidr-el2 isa-is-simobject remove-arm-cpsr-mode-miscreg arm-gem5-gic-ext 

Traceback (most recent call last):
  File "/usr/lib/python3.10/configparser.py", line 791, in get
    value = d[option]
  File "/usr/lib/python3.10/collections/__init__.py", line 986, in __getitem__
    return self.__missing__(key)            # support subclasses that define __missing__
  File "/usr/lib/python3.10/collections/__init__.py", line 978, in __missing__
    raise KeyError(key)
KeyError: 'regVal'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/harshilp/worktrees/gem5/./util/cpt_upgrader.py", line 363, in <module>
    process_file(cpt_file, **vars(args))
  File "/home/harshilp/worktrees/gem5/./util/cpt_upgrader.py", line 266, in process_file
    Upgrader.get(tag).update(cpt, tags)
  File "/home/harshilp/worktrees/gem5/./util/cpt_upgrader.py", line 152, in update
    self.upgrader(cpt)
  File "<string>", line 50, in upgrader
  File "/usr/lib/python3.10/configparser.py", line 794, in get
    raise NoOptionError(option, section)
configparser.NoOptionError: No option 'regVal' in section: 'board.processor.cores.core.isa'
```
I wanted to verify this as it also could be that the checkpoint it self is missing some stuff.

Navyera commented:
@Harshil2107 That is the behaviour I noticed locally as well... I filled in the missing keys by hand and then I was able to upgrade the checkpoint successfully. I thought that they were missing on purpose to make the checkpoint generic but maybe it is a problem...

Harshil2107 commented:
do you think it is possible to add the missing field in the hypervisor upgrader? I think completely re taking the checkpoint will pass the tests but not having a upgrader might break checkpoints that other users might have.

Navyera commented:
Unless I am misunderstanding what you mean, I do not think that hardcoding the addition of the missing keys in the hypervisor upgrader is a robust solution. 

For example, one of the missing keys is only requested by an x86 upgrader. It would make no sense to add these keys in a RISCV upgrader imo...

```
nick@tommy:~/CAL/hypervisor/gem5$ grep -R "regVal" ./util/cpt_upgraders/ 
./util/cpt_upgraders/isa-is-simobject.py:        "x86": ("regVal"),
./util/cpt_upgraders/x86-add-xcr0.py:                regVals = cpt.get(sec, "regVal")
./util/cpt_upgraders/x86-add-xcr0.py:                regVals = f"{regVals} 1"
./util/cpt_upgraders/x86-add-xcr0.py:                cpt.set(sec, "regVal", regVals)
./util/cpt_upgraders/x86-add-xcr0.py:            regVals = cpt.get(sec, "regVal")
./util/cpt_upgraders/x86-add-xcr0.py:            regVals = f"{regVals} 1"
./util/cpt_upgraders/x86-add-xcr0.py:            cpt.set(sec, "regVal", regVals)

```

Harshil2107 commented:
It looks like this is the issue, I will go through the existing upgraders and update them so only riscv ones are run to update the riscv checkpoints.

Harshil2107 commented:
It looks like the main issue is that the `isaName` field is missing from the checkpoint. This is an issue on multiple checkpoints in gem5 resources. I will see if I can write another updater that can automatically add this field correctly to the checkpoints.

rogerchang23424 commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -4835,41 +4804,93 @@ decode QUADRANT default Unknown::unknown() {
+                            xc->setMiscReg(status_idx, status);
+                            xc->tcBase()->getMMUPtr()->flushAll();
+
+                            if (!isV(xc)) {
+                                if (misa.rvh) {
+                                    if (hstatus.spv) {
+                                        setV(xc);
+                                    }
+                                    hstatus.spv = 0;
+                                    xc->setMiscReg(MISCREG_HSTATUS, hstatus);
+                                }
+
+                                status.mprv = 0; // 20211203
                             }
```
The part should handle before set status CSR

rogerchang23424 commented on src/arch/riscv/isa.cc:
```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
+RegVal
+ISA::readCSR(uint64_t csr)
+{
+    auto csr_it = CSRData.find(csr);
```
`getCSRDataMap()`? So that we can get customized CSR without panic

Ref: https://gem5-review.googlesource.com/c/public/gem5/+/62891

rogerchang23424 commented on src/arch/riscv/isa.cc:
```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
+void
+ISA::writeCSR(uint64_t csr, RegVal writeData)
+{
+    auto csr_it = CSRData.find(csr);
```
ditto

rogerchang23424 reviewed (commented):
Some cooment is the PR. Most of part is LGTM.

powerjg commented on src/arch/riscv/fault_codes.hh:
Needs a copyright statement.

powerjg commented on src/arch/riscv/RiscvISA.py:
```diff
@@ -107,7 +108,10 @@ class RiscvISA(BaseISA):
     privilege_mode_set = Param.PrivilegeModeSet(
-        "MSU",
+        "MSU",  # set MHSU to enable hypervisor (H-extension)
+        # O3 CPU is not supported in MHSU currently
```
Why does O3CPU not work? This is not a problem, but we need to have some detailed documentation.

Navyera replied:
Added some additional explanation

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -242,7 +275,7 @@ namespace RiscvISA
     [MISCREG_HPMCOUNTER31H]  = "HPMCOUNTER31H",
 
-    [MISCREG_FFLAGS_EXE]    = "FFLAGS_EXE",
+    [HELPER_FFLAGS_EXE] = "FFLAGS_EXE",
```
Why the change from MISCREG->HELPER?

Navyera replied:
imo it helps to have the invariance: MISCREG = PHYSREG, and the previous change breaks this.

The fact that OutOfOrder execution needs to buffer intermediate values before actually writing them to the physical register, does not mean that an actual implementation would necessarily add additional physical registers.

Hence, I designated these extra buffer registers as HELPER.

powerjg replied:
I don't know if I understand the need to change this variable name. I see your point, but it's changing the *meaning* of something that's used in other places in the code.

If you are strongly for keeping this change, then it's necessary to (1) add comments explaining the change in the code and (2) put this change in a separate commit so that the blame/log explains the change in semantics.

I know this seems like a lot to ask for a little thing, but I am thinking about the future when people come back to this code and wonder "what does "helper" mean, and why is this not MISCREG"

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -307,7 +339,7 @@ ISA::copyRegsFrom(ThreadContext *src)
 
     // Copying Misc Regs
-    for (int i = 0; i < NUM_PHYS_MISCREGS; i++)
+    for (int i = 0; i < NUM_MISCREGS; i++)
```
Why this change?

Navyera replied:
See above reply

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -705,7 +728,7 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
                     }
                 }
-
+                tc->getMMUPtr()->flushAll();
```
I assume it is in the spec that you have to flush the TLB when the PMP is updated?

Navyera replied:
yes, spike code does this as well

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
 }
 
+bool
+ISA::isV() const
```
Can we name this something more descriptive? Maybe "isInVirtualMode" or something like that?

Navyera replied:
Renamed to virtualizationEnabled

powerjg commented on src/arch/riscv/memflags.hh:
```diff
@@ -0,0 +1,26 @@
+    {
+        HLVX = 1ULL << 3,
+        FORCE_VIRT = 1ULL << 4,
+        LR = 1ULL << 5,
```
Can you add a comment to each of these which explains what "HLVX", "FORCE_VIRT", and "LR" means

Navyera replied:
done

powerjg commented on src/arch/riscv/mmu.hh:
```diff
@@ -76,10 +76,11 @@ class MMU : public BaseMMU
 
-    PrivilegeMode
-    getMemPriv(ThreadContext *tc, BaseMMU::Mode mode)
+    MemAccessInfo
+    getMemAccessInfo(ThreadContext *tc, BaseMMU::Mode mode)
     {
-        return static_cast<TLB*>(dtb)->getMemPriv(tc, mode);
```
Is this function no longer needed?

Navyera replied:
correct

powerjg commented on src/arch/riscv/regs/misc.hh:
```diff
@@ -246,22 +283,11 @@ enum MiscRegIndex
 
-    NUM_PHYS_MISCREGS,
-
-    MISCREG_MSTATUS = MISCREG_STATUS,
-    MISCREG_MIP = MISCREG_IP,
-    MISCREG_MIE = MISCREG_IE,
-    // This CSR shared the same space with MISCREG_FFLAGS
-    MISCREG_FFLAGS_EXE = NUM_PHYS_MISCREGS,
-    MISCREG_FCSR,
-    MISCREG_USTATUS,
-    MISCREG_UIP,
-    MISCREG_UIE,
-    MISCREG_SSTATUS,
-    MISCREG_SIP,
-    MISCREG_SIE,
```
Why is it OK to delete these?

powerjg commented on src/arch/riscv/tlb.cc:
```diff
@@ -310,56 +340,137 @@ TLB::doTranslate(const RequestPtr &req, ThreadContext *tc,
+        fault = NoFault;
+    }
+    else {
+
+
```
Extra newlines

powerjg commented on src/arch/riscv/tlb.cc:
```diff
@@ -310,56 +340,137 @@ TLB::doTranslate(const RequestPtr &req, ThreadContext *tc,
+
+    MISA misa = tc->readMiscReg(MISCREG_ISA);
+
+    bool special_access = false
```
What does "special" mean in this context?

Navyera replied:
added explanation in the code. special_access = bypass the TLB and don't insert

powerjg commented on src/arch/riscv/tlb.cc:
```diff
@@ -310,56 +340,137 @@ TLB::doTranslate(const RequestPtr &req, ThreadContext *tc,
 
-    Addr paddr = e->paddr << PageShift | (vaddr & mask(e->logBytes));
+    if (fault != NoFault)
+        return fault;
```
could `e` leak if this is hit? I.e., you don't `delete e`

Navyera replied:
Nice catch, added the delete

powerjg commented on src/arch/riscv/tlb.cc:
```diff
@@ -310,56 +340,137 @@ TLB::doTranslate(const RequestPtr &req, ThreadContext *tc,
+        fault = walker->start(tc, translation, req, mode);
+        if (translation != nullptr || fault != NoFault) {
+            delayed = true;
             return fault;
```
could `e` leak if this is hit? I.e., you don't `delete e`

Navyera replied:
Nice catch, added the delete

powerjg commented on util/cpt_upgraders/riscv-hext.py:
This file needs a copyright statement.

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -189,10 +197,44 @@ Walker::WalkerState::initState(ThreadContext * _tc,
+
+    entry.asid = satp.asid;
+    if (debug_condition()) {
+        DPRINTF(PageTableWalker, "======================================\n");
```
Is this absolutely necessary? I think prints like this end up spamming the output.

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -670,10 +1071,29 @@ Walker::WalkerState::retry()
+}
+
+bool
+Walker::WalkerState::debug_condition() {
```
This should be removed. If you want to print things in separately from the current debug flags, add a new one. You can also add compound debug flags. 

I can understand wanting to do this for a subset of addresses, but I don't think a one-off implementation in the ptw is the right way to do this.

Navyera replied:
removed this completely, as well as all usages.

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -242,91 +284,355 @@ Walker::startWalkWrapper()
+    glevel = SV39_LEVELS - 1;
+
+    // TODO THE TIMING PATH IS UNTESTED
+    if (timing)
+    {
```
The style is not right in this block of code. E.g.,

```suggestion
    if (timing) {
```

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -242,91 +284,355 @@ Walker::startWalkWrapper()
 Walker::WalkerState::stepWalk(PacketPtr &write)
 {
-    assert(state != Ready && state != Waiting);
+    int &curlevel = curstage == GSTAGE ? glevel : level;
```
Why is this a reference?

Navyera replied:
To change the actual object members when we write : curlevel--;

powerjg replied:
I think this syntax is confusing. What I remind my students often is that the purpose of code isn't for humans to communicate to computers, but for humans to communicate to other humans. The *intention* of this code is not clear to future humans who will come back and look at this code and try to understand it. 

Another way to look at it is that compilers optimize human-focused code so computers can understand it. Thus, the code you feed to the compiler should be optimized for readability, not to reduce the total number of characters.

powerjg replied:
Also, this should be `cur_level` since it is a local variable. Also, from reading the code more closely in the rest of my review, it is even more confusing that we are using a reference.

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -399,21 +715,55 @@ Walker::WalkerState::stepWalk(PacketPtr &write)
+                        // (GStage reaches here multiple times)
+                        if (walkType == OneStage || walkType == GstageOnly ||
+                            (walkType == TwoStage && curstage == FIRST_STAGE))
+                        {
```
Style comment: This is *way* too deep of nesting. I suggest adding a function.

Navyera replied:
Completely agree, but I did not want to extract out the logic, especially since all the previous steps are numbered inline with comments.

Still, we can refactor this if it is a problem.

powerjg replied:
I think this does need be refactored, along with other parts of this file. The problem is that if someone new comes to the code (or someone like me who is coming back to this code and is only familiar with the old code), they can't understand what the logical progression is. Since things are done in very long functions with deeply nested if/while statements, it is very hard for humans to understand the code. And, that's what this should be all about. Helping future humans understand the code. Also, I think that refactoring this code will be absolutely required when implementing timing accesses anyway.

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -485,8 +857,10 @@ Walker::WalkerState::stepWalk(PacketPtr &write)
 
-        DPRINTF(PageTableWalker,
-                "Loading level%d PTE from %#x\n", level, nextRead);
+        if (debug_condition()) DPRINTF(PageTableWalker,
+                "[%s] Get level%d PTE from %#x\n",
+                curstage == GSTAGE ? "GSTAGE" : "1stSTAGE",
+                curlevel, nextRead);
```
this really needs curly braces. it's quite confusing to read and could lead to bugs.

Navyera replied:
was removed anyway

powerjg reviewed (changes requested):
It would be great if you could add the tests that you used (e.g., comparing this to spike) to the gem5 test suite. 

Also, we need to update the RELEASE_NOTES and make it very, very clear that this only works for atomic mode and not for any timing CPU. (Not just O3, but also it doesn't work for timing or minor, right?)

Most of my comments below are about style. The page table walker file needs a lot of improvement to come up to our style standards.

Finally, with the staging branch being created next week, unfortunately, I think it's quite unlikely we can get this merged before we do the 24.1 release.

Harshil2107 commented:
I will retake the checkpoints after the changes are approved.

Navyera committed:
arch-riscv: Add copyright statements & memflag explanation

Navyera committed:
arch-riscv: remove ptw debug & rename isV & various fixes

Navyera committed:
arch-riscv: Merge mistakenly leftover changes

Navyera commented:
Hi @powerjg , I just rebased on top of current develop with changes addressing your comments. 
By the way, let me know if you prefer an alternative workflow. I am most familiar and confident with rebasing, but I can do it some other way, if there are benefits to alternative approaches.

@Harshil2107 could you please retake the checkpoints, so that the restore tests stop complaining?

powerjg commented on src/arch/riscv/RiscvISA.py:
```diff
@@ -107,7 +108,11 @@ class RiscvISA(BaseISA):
     privilege_mode_set = Param.PrivilegeModeSet(
-        "MSU",
+        "MSU",  # set MHSU to enable hypervisor (H-extension)
+        # All timing CPUs are NOT supported in MHSU currently
```
```suggestion
        # No timing CPUs are supported in MHSU currently
        # *ONLY THE ATOMIC CPU / ATOMIC MEMORY* is supported
```

Let me know if that is a misinterpretation. The comment needs to be a bit more explicit than it is now.

powerjg commented on src/arch/riscv/faults.cc:
```diff
@@ -137,6 +165,71 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
+                    default:
+                        break;
+                }
+            }
+            else {
```
```suggestion
            } else {
```

powerjg commented on src/arch/riscv/faults.cc:
```diff
@@ -137,6 +165,71 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
+                }
+                // Write changes to hstatus
+                tc->setMiscReg(MISCREG_HSTATUS, hstatus);
+            }
+            else if (prv == PRV_S &&
```
```suggestion
            } else if (prv == PRV_S &&
```

powerjg commented on src/arch/riscv/faults.cc:
```diff
@@ -137,6 +165,71 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
+                }
+                // Going to M-mode for handling, disable V if it's on
+                if (virtualizationEnabled(tc)) { resetV(tc); }
+            }
+            else if (prv == PRV_S &&
```
```suggestion
            } else if (prv == PRV_S &&
```

powerjg commented on src/arch/riscv/interrupts.cc:
```diff
@@ -215,20 +286,22 @@ void
-    ip = ip_ulong;
-    UNSERIALIZE_SCALAR(ie_ulong);
-    ie = ie_ulong;
+    UNSERIALIZE_SCALAR(ip_ulong);   ip = ip_ulong;
+    UNSERIALIZE_SCALAR(hvip_ulong); hvip = hvip_ulong;
+    UNSERIALIZE_SCALAR(ie_ulong);   ie = ie_ulong;
```
```suggestion
    UNSERIALIZE_SCALAR(ip_ulong);   
    ip = ip_ulong;
    UNSERIALIZE_SCALAR(hvip_ulong); 
    hvip = hvip_ulong;
    UNSERIALIZE_SCALAR(ie_ulong);   
    ie = ie_ulong;
```

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -565,20 +604,22 @@ ISA::readMiscReg(RegIndex idx)
+                                          STATUS_SUM_MASK;
+
+                // If at least one of these changed, flush
+                if ((oldstatus ^ status) & bits_of_interest) {
```
If each of these is a boolean variable, can you change this to use boolean logic instead of bitwise logic? (e.g., `&&` instead of `&`)

Navyera replied:
The values are whole registers in this case so the bitwise operation makes sense (if I am not mistaken)

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
+                    "SATP access with TVM enabled\n",
+                    machInst);
+        }
+    }
+    else if (csr == CSR_VSATP) {
```
```suggestion
    } else if (csr == CSR_VSATP) {
```

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
+                "VSATP access with hstatus.vtvm enabled",
+                machInst);
+        }
+    }
+    else if (csr == CSR_HGATP) {
```
```suggestion
    } else if (csr == CSR_HGATP) {
```

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -985,6 +1018,250 @@ ISA::resetThread()
+    return NoFault;
+}
+
+RegVal
+ISA::backdoorReadCSRAllBits(uint64_t csr) {
```
```suggestion
RegVal
ISA::backdoorReadCSRAllBits(uint64_t csr) 
{
```

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -995,6 +1272,93 @@ ISA::getFaultHandlerAddr(RegIndex idx, uint64_t cause, bool intr) const
+}
+
+// FPU status update function
+Fault updateFPUStatus(ExecContext *xc, ExtMachInst machInst, bool set_dirty) {
+
+
+    MISA misa = xc->readMiscReg(MISCREG_ISA);
```
```suggestion
Fault 
updateFPUStatus(ExecContext *xc, ExtMachInst machInst, bool set_dirty) 
{
    MISA misa = xc->readMiscReg(MISCREG_ISA);
```

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -995,6 +1272,93 @@ ISA::getFaultHandlerAddr(RegIndex idx, uint64_t cause, bool intr) const
+    return NoFault;
+}
+
+// VPU status update function
+Fault updateVPUStatus(
+    ExecContext *xc, ExtMachInst machInst, bool set_dirty, bool check_vill) {
```
Style

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -995,6 +1272,93 @@ ISA::getFaultHandlerAddr(RegIndex idx, uint64_t cause, bool intr) const
+    STATUS vsstatus = misa.rvh && virtualizationEnabled(xc) ?
+        xc->readMiscReg(MISCREG_VSSTATUS) : 0;
+
+    if (!misa.rvv || status.vs == VPUStatus::OFF ||
+        (misa.rvh && virtualizationEnabled(xc) &&
+         vsstatus.vs == VPUStatus::OFF))
+        return std::make_shared<IllegalInstFault>(
+            "RVV is disabled or VPU is off", machInst);
```
Must have curly braces since the inner block is multiple lines. However, IMO all `if` statements should always have curly braces. Again, I have spent many hours tracking down bugs because someone tried to save 2 characters. It ends up being not worth it in the end :)

powerjg commented on src/arch/riscv/tlb.cc:
```diff
@@ -248,50 +249,79 @@ TLB::remove(size_t idx)
-            DPRINTF(TLB, "PTE is not user accessible, raising PF\n");
-            fault = createPagefault(vaddr, mode);
+        if (priv == PRV_U && !pte.u) {
+            pf = true; DPRINTF(TLB, "PTE not user accessible, raising PF\n");
         }
-        else if (pmode == PrivilegeMode::PRV_S && pte.u && status.sum == 0) {
-            DPRINTF(TLB, "PTE is only user accessible, raising PF\n");
-            fault = createPagefault(vaddr, mode);
+        else if (priv == PRV_S && pte.u &&
+                (mode == BaseMMU::Execute || sum == 0)) {
+            pf = true; DPRINTF(TLB, "PTE only user accessible, raising PF\n");
         }
     }
 
-    return fault;
+    return pf ? createPagefault(vaddr, mode, gvaddr, gpf, virt) : NoFault;
 }
 
 Fault
-TLB::createPagefault(Addr vaddr, BaseMMU::Mode mode)
+TLB::createPagefault(Addr vaddr, BaseMMU::Mode mode,
+                     Addr gvaddr, bool gpf, bool virt)
 {
     ExceptionCode code;
     if (mode == BaseMMU::Read)
-        code = ExceptionCode::LOAD_PAGE;
+        code = gpf ? ExceptionCode::LOAD_GUEST_PAGE :
+                     ExceptionCode::LOAD_PAGE;
     else if (mode == BaseMMU::Write)
-        code = ExceptionCode::STORE_PAGE;
+        code = gpf ? ExceptionCode::STORE_GUEST_PAGE :
+                     ExceptionCode::STORE_PAGE;
     else
-        code = ExceptionCode::INST_PAGE;
-    return std::make_shared<AddressFault>(vaddr, code);
+        code = gpf ? ExceptionCode::INST_GUEST_PAGE :
+                     ExceptionCode::INST_PAGE;
+
+    if (gpf) { assert(virt); }
+    return std::make_shared<AddressFault>(vaddr, code, gvaddr, virt);
```
Curly braces!

powerjg commented on src/arch/riscv/tlb.cc:
```diff
@@ -400,6 +536,20 @@ TLB::translate(const RequestPtr &req, ThreadContext *tc,
 
+        // according to the RISC-V tests, negative physical addresses trigger
+        // an illegal address exception.
+        // TODO where is that written in the manual?
```
Do you want to fill in this "todo" before committing?

powerjg commented on util/cpt_upgraders/riscv-hext.py:
```diff
@@ -0,0 +1,106 @@
+                print(
+                    "MISCREG_* RVH registers already seem " "to be inserted."
+                )
+            else:
```
Do we need to be worried that we change the name from MISCREG_FFLAGS to HELPER_...?

powerjg commented on src/arch/riscv/pagetable_walker.cc:
```diff
@@ -242,91 +278,284 @@ Walker::startWalkWrapper()
+        // DO GSTAGE HERE IF NEEDED AND THEN DO PACKETS FOR *PTE
+        sendPackets();
+    }
+    else
```
Is it possible to split this out into another function. The `startWalk` function is now *very* complicated. (And it's renamed? Unfortunately, I no longer understand what the walker's logical algorithm is.)

powerjg commented on src/arch/riscv/isa/templates/vector_arith.isa:
The changes in these two files make me feel uncomfortable because we don't have any tests for them. There could be some subtle differences in the code blocks, but it looks like all of the changes were just copy-paste. How did you test that these changes have no effects on vector code?

Navyera replied:
Well, since there are no tests as you said I would consider the change untested indeed. We are not familiar with the vector arithmetic stuff so I do not think we can adequately test this code ourselves. 

The change was done to mirror the changes for fpuStatus (updateVPUStatus is similar to the pattern for updateFPUStatus). We simply saw a chance to avoid some duplication and improve readability. We can undo the change if it is deemed unsafe.

powerjg reviewed (commented):
It would be good if we could produce a `fatal` at initialization if we try to create a system with a timing-based CPU when enabling H-mode. However, it may be that we can only do it at runtime. Either way, the user should be informed with a very clear message that only atomic is supported when using the h-extension.

Lots of little things below. I think the biggest comment is that the changes to the page table walker are so complex that they really need to be refactored so future people can understand them (and to prepare them to be used in timing mode). Otherwise, there are lots of little style things (which, tbh, I didn't comment on every one of them). 

@rogerchang23424 I resolved some of your comments where it was clear that the authors made the updates, but it would be good if you could re-review. Also, please feel free to disagree with my comments. You're more of an expert on this code than I am.

Sorry @Navyera for taking so long to get back to this. My excuses are (1) This is a *HUGE* amount of code to review, so it takes having a large chunk of time to concentrate. (2) The release. (3) The holidays. Thanks for pinging me on it, and I'll do my best to help you get things merged as soon as we can.

rogerchang23424 commented:
I will review again once the PR is on the ToT

Navyera committed:
arch-riscv: Minor style and naming fixes

Navyera committed:
arch-riscv: refactor PTW logic to separate GStage logic

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Navyera commented:
Hi @rogerchang23424  @powerjg , I just rebased on top of the current develop.

I have hopefully addressed most of the style comments brought up by @powerjg and also attempted a refactor of the PTW logic to clearly separate GStage logic and not overcomplicate the old code (where possible). We are open to comments on the refactor.

powerjg commented on src/arch/riscv/interrupts.cc:
```diff
@@ -72,17 +73,48 @@ Interrupts::globalMask() const
+                INTERRUPT hideleg = tc->readMiscReg(MISCREG_HIDELEG);
+
+                mask.local = ~hideleg.local | ~mideleg.local;
+                if (status.uie)
+                    mask.local = mask.local | (hideleg.local & mideleg.local);
```
Use curly braces

powerjg commented on src/arch/riscv/interrupts.cc:
```diff
@@ -91,24 +123,61 @@ Interrupts::globalMask() const
+                INTERRUPT hideleg = tc->readMiscReg(MISCREG_HIDELEG);
+
+                mask.local = ~hideleg.local | ~mideleg.local;
+                if (status.sie)
+                    mask.local = mask.local | (hideleg.local & mideleg.local);
```
Curly brace

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -804,9 +810,15 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
                     new_val.mode = cur_val.mode;
+
+                // Flush TLB if satp changed as Spike does
+                if (new_val != cur_val) {
```
Let's look a little deeper here. If the current implementation is correct (i.e., if it functionally works without flushing the TLB), then we can probably elide this. It's also possible that the original gem5 code was wrong and we could have read stale translations.

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -827,6 +839,20 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
+                SATP cur_val = readMiscRegNoEffect(idx);
+                SATP new_val = val;
+
+                if (new_val.mode != AddrXlateMode::BARE &&
```
Curly braces, please :)

powerjg reviewed (commented):
Let's do one more quick round. Double check all of the comments, rebase (again, I'm sorry 😢) and then Roger can take one final pass.

Navyera committed:
arch-riscv: Use getCSRDataMap() instead of CSRData member

Navyera committed:
arch-riscv: Curly braces, various places

erin-le commented:
@Navyera: Hello, thank you for your contribution! Could you respond to the new set of comments when you have time?

erin-le commented:
@Navyera Hello, just checking in again. Could you give an update on the status of this PR on your end? Thank you!

Navyera commented:
Hi @erin-le, I am sorry for the late reply on this. We are currently rebasing and re-testing everything on top of the current develop branch. Hoping to sync with the reviewers so once I push the rebased commits the final review pass can take place.

We had paused work on this for a while, but we're now actively working to get it ready for merge as soon as possible.

Navyera committed:
arch-riscv: Fix interrupt checking functions

Navyera committed:
arch-riscv: Move status write to the end of sret

Navyera committed:
arch-riscv: Include SvNAPOT changes

Navyera committed:
arch-riscv: Use xc in read/write CSR for miscReg stats

Navyera commented:
@powerjg @erin-le @rogerchang23424 
Just completed and pushed the rebase. Hoping we can finalize things and put this on the staging branch.

Navyera committed:
arch-riscv: Elide unecessary TLB flushes

powerjg commented:
@erin-le can you create a place to track things that we want on the staging branch? This should be one of those things :)

BobbyRBruce commented:
@Navyera: Are you still working on this? We can't merge it with these tests failing

Navyera commented:
Hi @BobbyRBruce , the failing tests are due to checkpoints that must be updated with the cpt-upgrader we had provided.
There was a discussion about this starting [here](https://github.com/gem5/gem5/pull/1387#issuecomment-2402509187) and continuing in the following comments, but it never got resolved.

Harshil2107 committed:
tests: Update riscv checkpoint test to use updated checkpoint

Harshil2107 commented:
I have updated the checkpoint using the new  cpt upgrader in this PR. 

Here is the link to the updated checkpoint: https://resources.gem5.org/resources/riscv-hello-example-checkpoint?version=6.0.0

I have also pushed a commit to update the tests to use this new checkpoint.

erin-le commented:
@Navyera: Hello, it unfortunately looks like this PR is causing the [SST tests](https://github.com/gem5/gem5/actions/runs/15308939906/job/43068592271) to fail. The specific error message is as follows:
```
system.remote_gdb: Listening for connections on port 7000
build/RISCV/arch/riscv/linux/fs_workload.cc:53: info: Loading DTB file: m5out/device.dtb at address 0x87e00000
gem5Component-gem5_node-> init phase: 1
gem5Component-gem5_node-> init phase: 2
gem5Component-gem5_node-> init phase: 3
gem5Component-gem5_node-> init phase: 4
gem5Component-gem5_node-> init phase: 5
gem5Component-gem5_node->Component is being setup.
gem5.cc:295: info: Entering event queue @ 0.  Starting simulation...
sst: build/RISCV/arch/riscv/tlb.cc:335: gem5::Addr gem5::RiscvISA::TLB::hiddenTranslateWithTLB(gem5::Addr, uint16_t, gem5::Addr, gem5::BaseMMU::Mode): Assertion `e != nullptr' failed.
Program aborted at tick 327201500305
--- BEGIN LIBC BACKTRACE ---
```
To run the SST tests locally, I did the following:
1. Ran `docker run -u $UID:$GID --volume ./:/gem5 --rm -it ghcr.io/gem5/sst-env:latest` in my gem5 directory
2. Ran `scons build/RISCV/libgem5_opt.so --without-tcmalloc --duplicate-sources --ignore-style -j $(nproc)` in the `/gem5` directory in the Docker container
3. In the `/gem5/ext/sst` directory in the Docker container, ran `mv Makefile.linux Makefile`, `make -j $(nproc)`, and `sst --add-lib-path=./ sst/example.py`

The specific commands run by the SST tests can also be found in `gem5/.github/workflows/daily-tests.yaml`. Could you take a look and see if you can fix this failure? Thank you!

powerjg commented:
Note that the normal RISC-V boot also fails. The following command has the same error:

```sh
build/ALL/gem5.opt configs/example/gem5_library/riscv-fs.py
```

@Navyera We are aiming to make a release in ~1.5 weeks. We need to resolve this bug ASAP.

powerjg commented:
FYI, the problem occurs for load-reserved accesses. Why do we not lookup the TLB for load-reserved? I'm testing now to see if it works if I remove `memaccess.lr` from the check for `special_access` in `doTranslate`

powerjg commented:
I believe https://github.com/gem5/gem5/pull/2331 fixes this issue. @Navyera I would appreciate it if you could review.

Navyera commented:
@powerjg Thanks for identifying and fixing this so quickly!

We had not tested the boot with timing CPUs. The code expects to find that address in the TLB, but since we forced a TLB bypass the translation is missing, hence the error.

The reason we had modified the code in that way (forced the TLB bypass for LR) is the following code from the Spike ISA simulator:

```C
struct xlate_flags_t {
  const bool forced_virt : 1 {false};
  const bool hlvx : 1 {false};
  const bool lr : 1 {false};
  const bool ss_access : 1 {false};
  const bool clean_inval : 1 {false};

  bool is_special_access() const {
    return forced_virt || hlvx || lr || ss_access || clean_inval;
  }
};
...

...
    if (likely(!xlate_flags.is_special_access() && aligned && tlb_hit)) {
      res = *(target_endian<T>*)(tlb_data[vpn % TLB_ENTRIES].host_offset + addr);
    } else {
      load_slow_path(addr, sizeof(T), (uint8_t*)&res, xlate_flags);
    }
...

```

It is not yet clear to me why and in which cases LR would need a TLB bypass and I am not really familiar with the LR/SC pattern and its details in RISCV. We simply "trusted" Spike on this one.

We can look into the details of this further, but I think your fix addresses this for now.

powerjg commented:
@Navyera FYI, there's still a bug. Running a simple RISC-V boot sometimes hangs. If you run the gem5 script `configs/example/gem5_library/riscvmatched-fs.py --to-init` it never exits with this PR. Before this PR, it should exit after booting Linux.

I'm not sure what's wrong. I'm guessing it might have to do with the pseudo ops, but I don't have much evidence for that.

Note that booting ubuntu 24.04 seems to work as expected.

If we can't resolve this issue before the release next week I'm afraid we're going to have to revert this PR.

powerjg commented:
Looks like the problem is MinorCPU+20.04, btw. I'll test with MinorCPU+24.04 now.

Navyera commented:
Hi @powerjg , confirming the behaviour you're describing, we're also looking into possible causes/fixes. Did MinorCPU+24.04 work? I have only tested with the 20.04 image so far. That boot takes quite a while unfortunately...

powerjg commented:
> Did MinorCPU+24.04 work? 

Not with this PR applied. Without this PR, it works.

> That boot takes quite a while unfortunately...

Minor+20.04 only takes less than 10 minutes on my M4 Macbook.

Navyera commented:
@powerjg  I think we got it. #2355 should fix the problem. Let me know if the change works on your end as well.

P.S. I am really tempted to get an M4 MacBook 😄 The i9-14900K system I am using does that boot in ~24mins !!!
